zapcore/write_syncer.go

```diff
@@ -75,6 +77,53 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	bufferWriter *bufio.Writer
+	ws           WriteSyncer
+}
+
+// bufferSize sizes the buffer associated with each log file. It's large
+// so that log records can accumulate without the logging thread blocking
+// on disk I/O. The flushDaemon will block instead.
+const bufferSize = 256 * 1024
+
+// flushInterval means the default flush interval
+const flushInterval = 30 * time.Second
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// which is implemented in https://github.com/golang/glog.
+func Buffer(ws WriteSyncer) WriteSyncer {
+	if _, ok := ws.(*bufferWriterSyncer); ok {
+		// no need to layer on another buffer
+		return ws
+	}
+
+	// bufio is not goroutine safe, so add lock writer here
+	ws = Lock(&bufferWriterSyncer{
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+	})
+
+	// flush buffer every interval
+	// we do not need exit this goroutine explicitly
+	go func() {
+		for range time.NewTicker(flushInterval).C {
+			if err := ws.Sync(); err != nil {
```

Comment:
This is racy, simple repro:
```go
package main

import (
        "bufio"
        "fmt"
        "os"
)

func main() {
        w := bufio.NewWriter(os.Stdout)
        fmt.Fprintf(w, "hello world\n")
        go w.Flush()
        fmt.Fprintf(w, "hello 2\n")
}
```

Running this with `-race` causes data races.

Reply:
We get Lock for this WriteSync, so there's no race actually :-)
```
	// bufio is not goroutine safe, so add lock writer here
	ws = Lock(&bufferWriterSyncer{
		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
	})

	// flush buffer every interval
	// we do not need exit this goroutine explicitly
	go func() {
		for range time.NewTicker(flushInterval).C {
			if err := ws.Sync(); err != nil {
				return
			}
		}
	}()
```

Reply:
My bad, I missed the `Lock`, thanks!